// @antv/l7-core/es/services/layer/LayerService.js
import _initializerDefineProperty from "@babel/runtime/helpers/initializerDefineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";
import _initializerWarningHelper from "@babel/runtime/helpers/initializerWarningHelper";

var _dec, _dec2, _dec3, _class, _class2, _descriptor, _descriptor2;

import { inject, injectable } from 'inversify';
import { TYPES } from '../../types';
import Clock from '../../utils/clock';
var LayerService = (_dec = injectable(), _dec2 = inject(TYPES.IRendererService), _dec3 = inject(TYPES.IGlobalConfigService), _dec(_class = (_class2 = function () {
  function LayerService() {
    _classCallCheck(this, LayerService);

    _defineProperty(this, "clock", new Clock());

    _defineProperty(this, "alreadyInRendering", false);

    _defineProperty(this, "layers", []);

    _defineProperty(this, "layerRenderID", void 0);

    _defineProperty(this, "sceneInited", false);

    _defineProperty(this, "animateInstanceCount", 0);

    _initializerDefineProperty(this, "renderService", _descriptor, this);

    _initializerDefineProperty(this, "configService", _descriptor2, this);
  }

  _createClass(LayerService, [{
    key: "add",
    value: function add(layer) {
      if (this.sceneInited) {
        layer.init();
      }

      this.layers.push(layer);
    }
  }, {
    key: "initLayers",
    value: function initLayers() {
      this.sceneInited = true;
      this.layers.forEach(function (layer) {
        if (!layer.inited) {
          layer.init();
        }
      });
    }
  }, {
    key: "getLayers",
    value: function getLayers() {
      return this.layers;
    }
  }, {
    key: "getLayer",
    value: function getLayer(id) {
      return this.layers.find(function (layer) {
        return layer.id === id;
      });
    }
  }, {
    key: "getLayerByName",
    value: function getLayerByName(name) {
      return this.layers.find(function (layer) {
        return layer.name === name;
      });
    }
  }, {
    key: "remove",
    value: function remove(layer) {
      var layerIndex = this.layers.indexOf(layer);

      if (layerIndex > -1) {
        this.layers.splice(layerIndex, 1);
      }

      layer.emit('remove', null);
      layer.destroy();
      this.renderLayers();
    }
  }, {
    key: "removeAllLayers",
    value: function removeAllLayers() {
      this.destroy();
    }
  }, {
    key: "renderLayers",
    value: function renderLayers() {
      if (this.alreadyInRendering) {
        return;
      }

      this.alreadyInRendering = true;
      this.clear();
      this.updateRenderOrder();
      this.layers.filter(function (layer) {
        return layer.inited;
      }).filter(function (layer) {
        return layer.isVisible();
      }).forEach(function (layer) {
        layer.hooks.beforeRenderData.call();
        layer.hooks.beforeRender.call();
        layer.render();
        layer.hooks.afterRender.call();
      });
      this.alreadyInRendering = false;
    }
  }, {
    key: "updateRenderOrder",
    value: function updateRenderOrder() {
      this.layers.sort(function (pre, next) {
        return pre.zIndex - next.zIndex;
      });
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.layers.forEach(function (layer) {
        return layer.destroy();
      });
      this.layers = [];
      this.renderLayers();
    }
  }, {
    key: "startAnimate",
    value: function startAnimate() {
      if (this.animateInstanceCount++ === 0) {
        this.clock.start();
        this.runRender();
      }
    }
  }, {
    key: "stopAnimate",
    value: function stopAnimate() {
      if (--this.animateInstanceCount === 0) {
        this.stopRender();
        this.clock.stop();
      }
    }
  }, {
    key: "getOESTextureFloat",
    value: function getOESTextureFloat() {
      return this.renderService.extensionObject.OES_texture_float;
    }
  }, {
    key: "clear",
    value: function clear() {
      this.renderService.clear({
        color: [0, 0, 0, 0],
        depth: 1,
        stencil: 0,
        framebuffer: null
      });
    }
  }, {
    key: "runRender",
    value: function runRender() {
      this.renderLayers();
      this.layerRenderID = requestAnimationFrame(this.runRender.bind(this));
    }
  }, {
    key: "stopRender",
    value: function stopRender() {
      cancelAnimationFrame(this.layerRenderID);
    }
  }]);

  return LayerService;
}(), (_descriptor = _applyDecoratedDescriptor(_class2.prototype, "renderService", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor2 = _applyDecoratedDescriptor(_class2.prototype, "configService", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
export { LayerService as default };
